File: Rock/Model/AuditDetail.cs
// <copyright>
// Copyright by the Spark Development Network
//
// Licensed under the Rock Community License (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.rockrms.com/license
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// </copyright>
//
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Data.Entity.ModelConfiguration;
using System.Runtime.Serialization;

using Rock.Data;

namespace Rock.Model
{
    /// <summary>
    /// Represents an Audit Log entry that is created when an add/update/delete is performed against an <see cref="Rock.Data.IEntity"/> of an
    /// auditable <see cref="Rock.Model.EntityType"/>.
    /// </summary>
    [RockDomain( "Core" )]
    [NotAudited]
    [Table( "AuditDetail" )]
    [DataContract]
    [HideFromReporting]
    public partial class AuditDetail : Entity<AuditDetail>
    {

        #region EntityProperties

        /// <summary>
        /// Gets or sets the EntityTypeId for the <see cref="Rock.Model.EntityType"/> of entity that was modified. This property is required.
        /// </summary>
        /// <value>
        /// A <see cref="System.Int32"/> representing the EntityTypeId for the <see cref="Rock.Model.EntityType"/> of the entity that was modified.
        /// </value>
        [Required]
        [DataMember( IsRequired = true )]
        public int AuditId { get; set; }

        /// <summary>
        /// Gets or sets the Property.
        /// </summary>
        /// <value>
        /// Property.
        /// </value>
        [Required]
        [MaxLength( 100 )]
        [DataMember( IsRequired = true )]
        public string Property { get; set; }

        /// <summary>
        /// Gets or sets the Original Value.
        /// </summary>
        /// <value>
        /// Original Value.
        /// </value>
        [DataMember]
        public string OriginalValue { get; set; }

        /// <summary>
        /// Gets or sets the Current Value.
        /// </summary>
        /// <value>
        /// Current Value.
        /// </value>
        [DataMember]
        public string CurrentValue { get; set; }

        #endregion

        #region Virtual Properties

        /// <summary>
        /// Gets or sets the <see cref="Rock.Model.Audit"/> parent entity.
        /// </summary>
        /// <value>
        /// The <see cref="Rock.Model.Audit"/>
        /// </value>
        [LavaInclude]
        public virtual Model.Audit Audit { get; set; }

        #endregion

        #region Public Methods

        /// <summary>
        /// Returns a <see cref="System.String" /> that represents this instance.
        /// </summary>
        /// <returns>
        /// A <see cref="System.String" /> that represents this instance.
        /// </returns>
        public override string ToString()
        {
            return this.CurrentValue;
        }

        #endregion
    }

    #region Entity Configuration

    /// <summary>
    /// Entity Change Configuration class.
    /// </summary>
    public partial class AuditDetailConfiguration : EntityTypeConfiguration<AuditDetail>
    {
        /// <summary>
        /// Initializes a new instance of the <see cref="EntityTypeConfiguration" /> class.
        /// </summary>
        public AuditDetailConfiguration()
        {
            this.HasRequired( p => p.Audit ).WithMany( a => a.Details ).HasForeignKey( p => p.AuditId).WillCascadeOnDelete( true );
        }
    }

    #endregion

}

File: Rock.JavaScript.Obsidian/Framework/Elements/auditDetail.ts
// <copyright>
// Copyright by the Spark Development Network
//
// Licensed under the Rock Community License (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.rockrms.com/license
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// </copyright>
//

import { computed, defineComponent, PropType, ref } from "vue";
import { AuditDetail as AuditDetailViewModel } from "../ViewModels/auditDetail";

export default defineComponent({
    name: "AuditDetail",

    props: {
        modelValue: {
            type: Object as PropType<AuditDetailViewModel>,
            default: {}
        }
    },

    setup(props) {
        const id = computed(() => props.modelValue.id?.toString() ?? "");

        const guid = computed(() => props.modelValue.guid ?? "");

        const createdByPersonId = computed(() => props.modelValue.createdByPersonId);

        const createdByName = computed(() => props.modelValue.createdByName ?? "");

        const createdRelativeTime = computed(() => props.modelValue.createdRelativeTime);

        const modifiedByPersonId = computed(() => props.modelValue.modifiedByPersonId);

        const modifiedByName = computed(() => props.modelValue.modifiedByName ?? "");

        const modifiedRelativeTime = computed(() => props.modelValue.modifiedRelativeTime);

        const showId = ref(true);

        const getPersonLink = (personId: number): string => `/Person/${personId}`;

        /**
         * Event handler for when the Id/Guid label is clicked.
         */
        const onIdClick = (): void => {
            showId.value = !showId.value;
        };

        return {
            createdByName,
            createdByPersonId,
            createdRelativeTime,
            getPersonLink,
            guid,
            id,
            modifiedByName,
            modifiedByPersonId,
            modifiedRelativeTime,
            onIdClick,
            showId,
        };
    },

    template: `
<div class="row">
    <div class="col-md-4">
        <dl>
            <dt>Created By</dt>
            <dd>
                <a v-if="createdByPersonId" :href="getPersonLink(createdByPersonId)">{{ createdByName }}</a>
                <span v-else-if="createdByName">{{ createdByName }}</span>
                <small v-if="createdRelativeTime">&nbsp;({{ createdRelativeTime }})</small>
            </dd>
        </dl>
    </div>

    <div class="col-md-4">
        <dl>
            <dt>Modified By</dt>
            <dd>
                <a v-if="modifiedByPersonId" :href="getPersonLink(modifiedByPersonId)">{{ modifiedByName }}</a>
                <span v-else-if="createdByName">{{ modifiedByName }}</span>
                <small v-if="modifiedRelativeTime">&nbsp;({{ modifiedRelativeTime }})</small>
            </dd>
        </dl>
    </div>

    <div class="col-md-4">
        <dl v-if="showId">
            <dt @click.stop="onIdClick">Id</dt>
            <dd>{{ id }}</dd>
        </dl>
        <dl v-else>
            <dt @click.stop="onIdClick">Guid</dt>
            <dd>{{ guid }}</dd>
        </dl>
    </div>
</div>
`
});
